src/diagram/Timeline.ts

```diff
@@ -0,0 +1,90 @@
+import { Cancellable } from "../diagram/Diagram";
+import { Diagram, Memento } from "./Diagram";
+
+export class Snapshot<T extends Cancellable> {
+  constructor(readonly origin: Memento, readonly modifier: T) {}
+}
+
+export class Timeline<T extends Cancellable> {
+  private latest!: Memento;
+  private undoStack: Snapshot<T>[] = [];
+  private redoStack: T[] = [];
+
+  constructor(private diagram: Diagram) {
+    this.resetHistory();
+  }
+
+  /**
+   * a getter method that returns the diagram it stored
+   *
+   * @return Diagram
+   */
+  getDiagram(): Diagram | null {
```

Comment:
How about you also refactor the `getDiagram` to a getter function

Reply:
Changed